### PR TITLE
[hcl/en] Add guide for Terraform / HCL language

### PR DESCRIPTION
**Pull request by romaninsh**
- [x] I solemnly swear that this is all original content of which I am the original author
- [x] Pull request title is prepended with `[language/lang-code]`
- [x] Pull request touches only one file (or a set of logically related files with similar changes made)
- [x] Content changes are aimed at *intermediate to experienced programmers* (this is a poor format for explaining fundamental programming concepts)
 - [x] Yes, I have double-checked quotes and field names!

### CODE REVIEW

File: hcl.html.markdown [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    for b in aws_s3_bucket.coloured-bucket: 
+       trimsuffix(b.bucket_prefix, '-')
+         => b.bucket
+   }
+}
+
+
+```
+
+I will contribute more to this guide when I have time
```
Comment (mxschumacher) [inline hcl.html.markdown:249]: I don't think this line is necessary :)

File: hcl.html.markdown [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+// Modules can declare outputs, that can be optionally referenced
+// (see above), typically outputs appear at the bottom of the file or
+// in "outputs.tf".
+output "knowledge" {
+  value = "types so far, more to come"
+}
+```
+
+Terraform exist for managing "resources". A could be anything as long as it
```
Comment (mxschumacher) [inline hcl.html.markdown:126]: exists

File: hcl.html.markdown [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+// Top-level HCL file will interactively ask user values for the variables
+// which do not have default value
+variable "ready" {
+  description = "Ready to learn?"
+  type = bool
+  // default = true
+}
+
+// Module block consults a specified folder for *.tf files, would
+// effectively prefix all resources id's with "module.learn-basics."
```
Comment (mxschumacher) [inline hcl.html.markdown:29]: it is just ids in this case, no apostrophe: http://www.sussex.ac.uk/informatics/punctuation/apostrophe/possessives

File: hcl.html.markdown [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+```hcl-terraform
+variable "yaml-data" {
+
+  // config is sourced from .yaml file, so technically it is a
+  // map(any), but we can narrow down type like this:
+  type = map(string)
+}
+
+// You do not need to explicitly define providers, they all have reasonable
+// defaults with environment variables. Using resource that relies on
```
Comment (mxschumacher) [inline hcl.html.markdown:144]: Using **a** resource

File: hcl.html.markdown [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+// you can create intermediate products from other variables, modules,
+// and functions.
+
+locals {
+  ready = var.ready ? "yes": "no"
+
+  yaml = yamldecode(file("${path.module}/file-in-current-folder.yaml"))
+}
+
+// locals block can be defined multiple times, but all variable, resource
```
Comment (mxschumacher) [inline hcl.html.markdown:106]: variable-, resource- and local

File: hcl.html.markdown [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+variable "favourites" {
+  type: set
+  default = ["red", "blue"]
+}
+
+// When type is not specified or is mixed of scalars they will be
+// converted to strings.
+
+
+// IDEs (such as IntelliJ) syntax highlighting
```
Comment (mxschumacher) [inline hcl.html.markdown:90]: while I like and use IDEs from Jetbrains, I don't think it makes sense to include this reference in the LXIYM guide

File: hcl.html.markdown [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+## learn-basics
+
+```hcl-terraform
+
+// Variables are not automatically passed into modules
+// and can be typeless.
+variable "ready" {
+}
+
+// It is good practice to define type though. There are 3 primitive types
```
Comment (mxschumacher) [inline hcl.html.markdown:50]: It is a good practice to define a type

**Comment by mxschumacher** [comment]
@romaninsh Thank you very much for this contribution !! There are some tiny issues I flagged, maybe you can look them over

**Comment by romaninsh** [comment]
@mxschumacher - Apologies for not responding for a long time. I have now completed the guide, upgraded to include newly released 0.13 version language features and addressed all of the notes too. Please review again and let me know if anything else needs fixing.

**Comment by olafkotur** [comment]
Hey, do you know if this can be reviewed again? @mxschumacher

**Comment by romans-ovo** [comment]
@mribeirodantas - any chance we could merge this? I keep sending people to my PR to read the guide :D I'm happy to clean up the guide if it needs anything.

**Comment by nbehrnd** [comment]
@romans-ovo May you please check the .md in preparation with markdownlint for consistency?

Some time ago, I noticed there are entries of this compilation which do not fit well the website's presentation managed by a style file. This is an issue others equally observed; it may be resolved if lines of the .md file do not exceed the 80 characters per line threshold .and. an adjustment of the web browser's font parameters ([see here](https://github.com/adambard/learnxinyminutes-docs/issues/2611)).

Thus, I would like to invite you to check the .md in preparation, e.g., with the freely available open source checker [markdownlint](https://github.com/markdownlint/markdownlint).  Matching the threshold requested here in learnxinyminutes is the one the linter uses as (adjustable) default; the tool quickly reports this  as `error 13` back to the CLI, including a note which line should be revised.

If present, the tool equally will identify other issues which -- except for the header of the .md defining language, contributors, translators, filename -- may be improved breaking longer commands and comments into multiple lines, editing spaces/words/indentations.  After application to an entry e.g., about AWK, I noticed if the web page is print to .pdf (simplified representation, as offered by Firefox), the web browser no longer shrinks the characters in size this much to squeeze overly long lines into a physical line, too.

At the end of the .md file, long entries with cross references to other resources / web pages seem to remain functional by substitution of a pattern like 

``` markdown
## Further Reading

+ The current standard 2017 of language xy is documented in [the reference](the reference's address) especially outlining detail A and B.
+ In addition ...
```

by a "poly-line" pattern

``` markdown
## Further Reading

+ The current standard 2017 of language xy is documented in
  [the reference](the reference's address)
  especially outlining detail A and B.
+ In addition ...
```

because a new paragraph would require a new *blank line*, not just starting a new line.

It usually is mix of issues on may resolve quickly with regexes (e.g., `sed -i "s/ *$//" example.markdown.html` to trim trailing spaces), and others which require manual intervention which either resolve the issues for good, or at least improve the .md file's format.

**Comment by verhovsky** [comment]
@romans-ovo @romaninsh I have proofread the guide and I'm going to merge it. If you ever get a chance, can you please proof-read it again and update it. It talks about Terraform 0.13 but these days Terraform is on version 1.8.3.

**Pull request merged**